app/scripts/lib/rpc-method-middleware/handlers/ethereum-chain-utils.js

```diff
@@ -0,0 +1,241 @@
+  }
+
+  const _chainId = validateChainId(chainId, end);
+  if (!rpcUrls || !Array.isArray(rpcUrls) || rpcUrls.length === 0) {
+    throw ethErrors.rpc.invalidParams({
+      message: `Expected an array with at least one valid string HTTPS url 'rpcUrls', Received:\n${rpcUrls}`,
+    });
+  }
+
+  const isLocalhostOrHttps = (urlString) => {
+    const url = getValidUrl(urlString);
+    return (
+      url !== null &&
+      (url.hostname === 'localhost' ||
+        url.hostname === '127.0.0.1' ||
+        url.protocol === 'https:')
+    );
+  };
+
+  const firstValidRPCUrl = rpcUrls.find((rpcUrl) => isLocalhostOrHttps(rpcUrl));
+  const firstValidBlockExplorerUrl =
+    blockExplorerUrls !== null && Array.isArray(blockExplorerUrls)
+      ? blockExplorerUrls.find((blockExplorerUrl) =>
+          isLocalhostOrHttps(blockExplorerUrl),
+        )
+      : null;
+
+  if (!firstValidRPCUrl) {
+    throw ethErrors.rpc.invalidParams({
+      message: `Expected an array with at least one valid string HTTPS url 'rpcUrls', Received:\n${rpcUrls}`,
+    });
+  }
+
+  if (blockExplorerUrls !== null && !firstValidBlockExplorerUrl) {
+    throw ethErrors.rpc.invalidParams({
+      message: `Expected null or array with at least one valid string HTTPS URL 'blockExplorerUrl'. Received: ${blockExplorerUrls}`,
+    });
+  }
+
+  if (typeof chainName !== 'string' || !chainName) {
+    throw ethErrors.rpc.invalidParams({
+      message: `Expected non-empty string 'chainName'. Received:\n${chainName}`,
+    });
+  }
+
+  const _chainName =
+    chainName.length > 100 ? chainName.substring(0, 100) : chainName;
+
+  if (nativeCurrency !== null) {
+    if (typeof nativeCurrency !== 'object' || Array.isArray(nativeCurrency)) {
+      throw ethErrors.rpc.invalidParams({
+        message: `Expected null or object 'nativeCurrency'. Received:\n${nativeCurrency}`,
+      });
+    }
+    if (nativeCurrency.decimals !== 18) {
+      throw ethErrors.rpc.invalidParams({
+        message: `Expected the number 18 for 'nativeCurrency.decimals' when 'nativeCurrency' is provided. Received: ${nativeCurrency.decimals}`,
+      });
+    }
+
+    if (!nativeCurrency.symbol || typeof nativeCurrency.symbol !== 'string') {
+      throw ethErrors.rpc.invalidParams({
+        message: `Expected a string 'nativeCurrency.symbol'. Received: ${nativeCurrency.symbol}`,
+      });
+    }
+  }
+
+  const ticker = nativeCurrency?.symbol || UNKNOWN_TICKER_SYMBOL;
+  if (
+    ticker !== UNKNOWN_TICKER_SYMBOL &&
+    (typeof ticker !== 'string' || ticker.length < 2 || ticker.length > 6)
+  ) {
+    throw ethErrors.rpc.invalidParams({
+      message: `Expected 2-6 character string 'nativeCurrency.symbol'. Received:\n${ticker}`,
+    });
+  }
+
+  return {
+    chainId: _chainId,
+    chainName: _chainName,
+    firstValidBlockExplorerUrl,
+    firstValidRPCUrl,
+    ticker,
+  };
+}
+
+export async function switchChain(
+  res,
+  end,
+  origin,
+  chainId,
+  requestData,
+  networkConfigurationId,
```

Comment:
Should this be the `networkClientId`, not the `networkConfigurationId`? They are equivalent at the moment, but following https://github.com/MetaMask/core/pull/4286, network configurations won't have IDs (and `setActiveNetwork` is documented to take a `networkClientId`, anyway).

Reply:
Yes lets make it `networkClientId` for furture clarity, but more generally as you're pointing out in all of your comments we will have to retrieve and handle this value differently after we modify the `networkConfiguration` system. I'm actually a bit surprised that we'll still have `networkClientId`'s after your work but I haven't looked into exactly how you're going about it yet. Will take a peek now.

Reply:
updated here: https://github.com/MetaMask/metamask-extension/pull/24415/commits/2619ccb6f02998350651a273e4abb7869599e58e